Improve performance and binary size of til::enumset

lhecker opened the pull request:
This commit approximately doubles the performance of til::enumset
and reduces it's binary footprint by approximately 1kB.
Most of the binary size can be attributed to exception handling.

Unfortunately this commit removes assertions that the given values are less than
the number of bits in the `underlying_type`. However I believe this to be a good
trade-off as the tests previously only happened at runtime, while tests at
compile time would be highly preferable. Such tests are technically possible,
however MSVC fails to compile (valid) `static_assert`s containing
`static_cast`s over a parameter pack at the time of writing.
With future MSVC versions such checks can be added to this class.

This change was initially discussed in #10492, but was forgotten to
be considered before it was merged. Since the work was already done,
this commit re-introduces the optimization. It's free!

## PR Checklist
* [x] I work here.
* [x] Tests added/passed

## Validation Steps Performed
* Run `printf '\e]8;;http://example.com\e\\This is a link\e]8;;\e\\\n'` in WSL
* A wild dotted line appears ✔️

DHowett commented:
@msftbot make sure @j4james signs off

zadjii-msft reviewed (approved):
(static analysis build is exploded fyi)

j4james commented on src/inc/til/enumset.h:
```diff
@@ -3,130 +3,134 @@
+        constexpr enumset& set(const T pos, const bool val) noexcept
         {
-            _base::set(static_cast<size_t>(pos), val);
+            if (val)
+            {
+                set(pos);
+            }
+            else
+            {
+                reset(pos);
+            }
```
Would it be worthwhile trying to avoid the branch here? Something like:
```
reset(pos);
_data |= (underlying_type{ val } << static_cast<underlying_type>(pos));
```

lhecker replied:
I've did some benchmarks and settled on basically
```cpp
const auto mask = 1 << pos;
_data = (_data & ~mask) | (-val & mask);
```

j4james replied:
Even better. I love this.

DHowett replied:
sh*t, that's clever

j4james commented on src/til/ut_til/EnumSetTests.cpp:
```diff
@@ -53,39 +53,39 @@ class EnumSetTests
-        VERIFY_ARE_EQUAL(0b00000u, flags.to_ulong());
+        VERIFY_ARE_EQUAL(0b00000u, flags.bits());
 
         Log::Comment(L"Set bit 2 to true");
         flags.set(Flags::Two);
```
The point of this test case was originally to validate the single position `set` method with a `val` of true. With the new method naming this is going to use the multiple position method instead. To get the intended coverage, I'd suggest changing at least one of these `set` calls to:
```
flags.set(Flags::Two, true);
```

j4james commented on src/til/ut_til/EnumSetTests.cpp:
```diff
@@ -53,39 +53,39 @@ class EnumSetTests
-        VERIFY_ARE_EQUAL(0b10100u, flags.to_ulong());
+        VERIFY_ARE_EQUAL(0b10100u, flags.bits());
 
         Log::Comment(L"Set bit 0 to true");
         flags.set(Flags::Zero);
```
As above, technically this should now be:
```
flags.set(Flags::Zero, true);
```

lhecker replied:
I've changed this one and left the other as it is.

j4james commented on src/til/ut_til/EnumSetTests.cpp:
```diff
@@ -53,39 +53,39 @@ class EnumSetTests
 
         Log::Comment(L"Set bit 0 to false, leaving 4 true");
-        flags.set(Flags::Zero, false);
-        VERIFY_ARE_EQUAL(0b10000u, flags.to_ulong());
+        flags.reset(Flags::Zero);
```
Similar to the cases above, the the point of this test was to validate the single position `set` method with a `val` of false. To get the intended coverage, I'd suggest changing this back to:
```
flags.set(Flags::Zero, false);
```

j4james commented on src/til/ut_til/EnumSetTests.cpp:
```diff
@@ -101,15 +101,13 @@ class EnumSetTests
-        Log::Comment(L"Test bit 3 and 4 with the array operator");
-        VERIFY_IS_FALSE(flags[Flags::Three]);
-        VERIFY_IS_TRUE(flags[Flags::Four]);
+        Log::Comment(L"Test bits with the any method");
+        VERIFY_IS_FALSE(flags.any(Flags::One));
+        VERIFY_IS_TRUE(flags.any(Flags::Two));
+        VERIFY_IS_FALSE(flags.any(Flags::Three));
+        VERIFY_IS_TRUE(flags.any(Flags::Four));
```
The point of these test cases were to validate the `test` method and the array ops. The array ops aren't used anymore, but we still need coverage for the `test` method, so maybe just make them all `test` calls. The `any` method is already covered in the test cases further down.

lhecker replied:
Ah I forgot to revert those tests...

j4james reviewed (changes requested):
This is awesome, but I'd just like to make sure we still have full code coverage in the unit tests. The `set` method optimisation is up to you.

lhecker commented:
Sorry all, I've force-pushed the latest commit. I hope I addressed all concerns. As far as I can see the tests should cover all functions now. (For some reason I can't create a code coverage analysis in VS?)

DHowett commented:
Thank you for keeping the general shape of the API. :smile:

j4james commented on src/inc/til/enumset.h:
```diff
@@ -3,130 +3,136 @@
 
         // Method Description:
         // - Returns the value of the bit at the given position.
         //   Throws std::out_of_range if it is not a valid position
         //   in the bitset.
```
Just noticed this comment no longer applies.

j4james reviewed (approved):
I'm thrilled with this. Just noticed the minor comment issue, but that's no big deal.

DHowett reviewed (approved):
I remain surprised that `set(3, 1)` doesn't ~magically~ horrifically fall into the `set(T, bool)` overload. I'll trust it, _for now._
